src/macro.c

```diff
@@ -106,3 +106,47 @@ executemacro(int f, int n)
 	inmacro = FALSE;
 	return (TRUE);
 }
+
+int
+applymacro(int f, int n)
+{
+	if (curwp->w_markp == NULL) {
+		dobeep();
+		ewprintf("No mark set in this window");
+		return (FALSE);
+	}
+
+    /* (save-mark-and-excursion) save the state of the "." and mark */
+    struct line *odotp = curwp->w_dotp, *omarkp = curwp->w_markp;
+    int odoto = curwp->w_doto, odotline = curwp->w_dotline;
+    int omarko = curwp->w_marko, omarkline = curwp->w_markline;
+
+    if (curwp->w_dotline > curwp->w_markline) {
+        swapmark(FFRAND, 0);
+    }
+
+	struct line  *tmarkp = curwp->w_markp;
+	int tmarko = curwp->w_marko;
+    int tmarkline = curwp->w_markline;
+
+    while (curwp->w_dotline < curwp->w_markline) {
+        gotobol(FFRAND, 1);
+        executemacro(FFRAND, 1);
+        forwline(FFRAND, 1);
+
+        /* restore mark that may have been altered by the executed macro */
+        curwp->w_markp = tmarkp;
+        curwp->w_marko = tmarko;
+        curwp->w_markline = tmarkline;
+    }
+
+    /* (save-mark-and-excursion) restore the state of the "." and mark */
+    curwp->w_dotp = odotp;
+    curwp->w_doto = odoto;
+    curwp->w_dotline = odotline;
+    curwp->w_markp = omarkp;
+    curwp->w_marko = omarko;
+    curwp->w_markline = omarkline;
+
+    return (TRUE);
+}
```

Comment:
This function needs to be reformatted to match the Mg coding style:

 1. tabs not spaces for indent, see other functions in the same file
 2. variable declarations at start of function
 3. variable assignments below declarations

For (2), make sure you separate type and variable name with a tab.